### ISSUE TITLE
Dode

### ISSUE THREAD

**Issue by KarlikWin** [KarlikWin/Psi_in_time#1]
import matplotlib.pyplot as plt
import numpy as np

import pandas as pd
import seaborn as sns
from sklearn.metrics import roc_curve, roc_auc_score

from scipy.stats import entropy
from scipy.spatial.distance import jensenshannon
from pandas.api.types import is_numeric_dtype

def psi_in_time(data, data_expected, pred, col_month, q=10, missing_value=0.0001, plot=True, show=True, output_folder=None):
    """Calculates PSI (Population Stability Index) of a variable by month. Reference data set (where the default distribution of the variable is taken from) is passed as one of the arguments.

    Args:
        data (pd.DataFrame): data to calculate PSI on
        data_expected (pd.DataFrame): reference data where reference distribution is taken from
        pred (str): name of variable to calculate PSI of
        col_month (str): name of column with number of month (PSI is calcualted for each month separately)
        q (int, optional): Number of categories to split variable pred to during the PSI calculation. Defaults to 10.
        missing_value (float, optional): Value to fill missings with. Defaults to 0.0001.
        plot (bool, optional): Plot PSI chart? Defaults to True.
        show (bool, optional): Show the plotted chart on output? Defaults to True.
        output_folder (str, optional): Where to save the results. Defaults to None.

    Returns:
        pd.DataFrame: table with PSI values
    """

    data_stability_expected = pd.DataFrame()
    data_stability = pd.DataFrame()

    if is_numeric_dtype(data[pred]):
        bins = np.unique(np.percentile(data_expected.loc[data_expected[pred].notnull(), pred], np.linspace(0, 100, q+1)))
        bins = [-np.inf]+sorted(list(bins))[1:-1]+[np.inf]
        data_stability_expected[pred] = pd.cut(data_expected[pred], bins=bins).values.add_categories('missing').fillna(value='missing')
        data_stability[pred] = pd.cut(data[pred], bins=bins).values.add_categories('missing').fillna(value='missing')
    else:
        bins = data[pred].unique()
        data_stability_expected[pred] = data_expected[pred].astype('category').values.add_categories('missing').fillna(value='missing')
        data_stability[pred] = data[pred].astype('category').values.add_categories('missing').fillna(value='missing')

    data_stability_expected = data_stability_expected.set_index(data_expected.index)
    data_stability_expected['expected'] = data_stability_expected.index
    data_stability_expected = data_stability_expected.groupby(pred).count().T
    data_stability_expected.columns = list(data_stability_expected)
    data_stability_expected['All'] = data_stability_expected.sum(axis=1)
    data_stability_expected
    
    data_stability = data_stability.set_index(data.index)
    data_stability['index'] = data_stability.index
    data_stability[col_month] = data[col_month]    

    pivot = pd.pivot_table(data_stability,
                           values='index', 
                           index=col_month, 
                           columns=pred, 
                           aggfunc='count', 
                           margins=True, 
                           dropna=False, 
                           margins_name='All')    

    pivot = pivot.append(data_stability_expected)
    pivot = pivot.apply(lambda x: x/x['All'], axis=1)
    pivot.loc['expected',:] = pivot.loc['expected',:].replace(0, missing_value).fillna(missing_value)
    pivot = pivot.drop('All', axis=1)
    pivot = pivot.drop('All', axis=0)

    psi = ((pivot-pivot.loc['expected',:]) * np.log(pivot/pivot.loc['expected',:])).sum(axis=1)
    # psi2 = pivot.apply(lambda x: entropy(x.fillna(missing_value), pivot.loc['expected',:]) + entropy(pivot.loc['expected',:], x.fillna(missing_value)), axis=1)
    bhattacharyya_metric = np.sqrt(1-np.sqrt(pivot*pivot.loc['expected',:]).sum(axis=1))
    jensenshannon_metric = pivot.apply(lambda x: jensenshannon(x.fillna(0), pivot.loc['expected',:]), axis=1)

    pivot['PSI'] = psi
    pivot['bhattacharyya'] = bhattacharyya_metric
    pivot['jensenshannon'] = jensenshannon_metric

    if plot and len(pivot) > 0:
        
        plt.figure(figsize = (10,6))
        plt.title(pred)
        plt.plot(pivot.drop(['expected'])['PSI'], color='b', label='PSI')
        plt.plot(pivot.drop(['expected'])['bhattacharyya'], color='r', label='Bhattacharyya')
        plt.plot(pivot.drop(['expected'])['jensenshannon'], color='y', label='Jensen-Shannon')
        plt.axhline(0.1, linestyle='--', color='y')
        plt.axhline(0.25, linestyle='--', color='r')
        plt.axhline(0.25, linestyle='--', color='b')
        # plt.xticks(pivot.drop(['expected']).index, [col.strftime('%Y-%m') for col in pivot.drop(['expected']).index], rotation=90)
        plt.xticks(pivot.drop(['expected']).index, [col for col in pivot.drop(['expected']).index], rotation=90)
        plt.legend(loc='best')
        
   
        if output_folder is not None:
            plt.savefig(output_folder+'/psi_'+str(pred)+'.png', format='png', dpi=72, bbox_inches='tight')
        if show:
            plt.show()   
        
        plt.close() 
    
    return pivot
